### ISSUE TITLE
Dense linear algebra functions need encapsulations for matrices and vectors

### ISSUE THREAD

**Issue by mhoemmen** [oneapi-src/oneMKL#23]
# Summary

oneMKL's dense linear algebra functions should provide or use classes that encapsulate matrices and vectors, instead of taking vectors and matrices as 1-D `sycl::buffer` or raw pointers.  This would improve memory safety and usability, and make the interface more idiomatically C++.  It would also better align with the various linear algebra proposals currently being considered for the C++ Standard Library.

# Problem statement

## 1-D `sycl::buffer` does not correctly encapsulate matrices or vectors

The current oneMKL dense BLAS interface takes vectors and matrices in two different ways:

1. as 1-D `sycl::buffer` (e.g., `sycl::buffer<double, 1>`); and,
2. as raw pointers (e.g., `double*`).

Both overloads take the dimensions and strides as separate, integer arguments.  Using raw pointers has the same memory safety issues as the C BLAS interface; for discussion, see [P1674](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1674r0.md).  Using 1-D `sycl::buffer`, with separate integer dimension and stride arguments, has the following issues:

1. It's as memory unsafe as the C BLAS interface, but more verbose.
2. It discards dimension(s) that `sycl::buffer` already stores, in favor of extra integer arguments that might be incorrect.
3. Matrices are 2-D objects, and `sycl::buffer<T, 2>` exists, yet the interface takes matrices as `sycl::buffer<T, 1>`.
4. "Batched" interfaces like [`gemm_batch`](https://spec.oneapi.com/versions/latest/elements/oneMKL/source/domains/blas/gemm_batch.html) compound the issue by adding another dimension.

Using `sycl::buffer` directly as a matrix or vector interface would offer less functionality than the current oneMKL interface.  This is because `sycl::buffer` always expresses contiguous memory, but (C and Fortran) BLAS functions can work with strided, possibly noncontiguous memory.  There is a way  to create a "sub-buffer" of an existing `sycl::buffer`, but the sub-buffer must also be contiguous.  
(See Chapter 4.7.2 of the [SYCL 1.2.1 spec](https://www.khronos.org/registry/SYCL/specs/sycl-1.2.1.pdf).)  The proposed oneMKL interface already accepts strided, possibly noncontiguous memory, just like the C or Fortran BLAS.

## `basic_mdspan` could replace the raw pointers interface, but not the `sycl::buffer` interface

It would be tempting to use [`basic_mdspan`](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p0009r10.html) in place of `sycl::buffer`.  That would let callers express all the different strided or contiguous matrix or vector layouts that the C BLAS can already express.  It would also make this proposal nearly a subset of [a pending C++ Standard Library proposal](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1673r2.pdf).

This would work perfectly well for the raw pointers interface, but it would not work for the `sycl::buffer` interface.  The problem is that `basic_mdspan` is a "view" in the [C++ Standard Library sense](http://eel.is/c++draft/views).  Views do not own their storage.  This means that views can't (or shouldn't really) do the things that SYCL needs to do with buffers, such as possibly allocate temporary storage on host or device, or track data dependencies.

# Preferred solution

## Replace the raw pointer interface with a `basic_mdspan` interface

Using `basic_mdspan` instead of raw pointers would be more expressive, easier to use, and less error prone.  It would be no less accessible from other programming languages than oneMKL's current interface.  (The current interface is not an `extern "C"` interface; it uses namespaces and class references.)

## Consider `basic_mdarray` as a `sycl::buffer` wrapper

Please consider whether [`basic_mdarray`](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1684r0.pdf), the container variant of `basic_mdspan`, could correctly represent matrices or vectors backed by `sycl::buffer` storage.  (Note how oneMKL's functions all take `sycl::buffer` by reference.  On the other hand, `sycl::buffer` is reference counted, while `basic_mdarray`'s copy construction and copy assignment deep-copy just like `std::vector` and the other C++ Standard Library containers.)  If not, please consider whether `basic_mdarray` could be changed to fix this, or whether some other container or buffer type would be appropriate.  If the latter, perhaps this type could benefit from standardization, in SYCL and/or in the C++ Standard Library.

**Comment by jasukhar** [comment]
Thank @mhoemmen for very well described proposal. We will take a look what can be done in this area and get back to you.

**Comment by mhoemmen** [comment]
Another option could be just to drop the interface that takes `sycl::buffer`.  New users would be likely to try 2-D buffers, then wonder why their code doesn't compile.  More importantly, if you offer a matrix extension of `sycl::buffer`, and if your matrix class can handle strided noncontiguous data (as it needs to, in order to support current BLAS and LAPACK functionality), then you'll find yourself needing to track data dependencies for submatrices that might overlap.  Callers who can write nontrivial matrix algorithms that use submatrices are likely to know how to track data dependencies themselves.  You'll have this problem even for BLAS 1 (vector) functions, because they take one or more stride (e.g., `INCX`) arguments that might not be 1.

**Comment by petercad** [comment]
Hi Mark, thanks for putting this request together! I agree with you completely that `sycl::buffer` isn’t a reasonable vector/matrix abstraction and we’re really using it here as a smart pointer of sorts. Our presumption, in fact, was that most users would prefer the pointer-based API, but buffers were the only option in the existing SYCL 1.2.1 spec.

Our goal for v1.0 of the oneMKL specification is to drive for a certain level of simplicity and basic consistency with existing C++ BLAS APIs (in particular BLAS++) to ease the migration of existing HPC applications to DPC++ across a broad range of architectures. We feel that a more elegant and feature rich C++-based BLAS proposal would require extensive coordination with and buy-in from University of Tennessee Netlib owners. Also, relying on mdspan or mdarray brings external dependencies since they haven’t reached STL yet.

That said, my thought is: let’s keep this idea on the table for the future, especially as `mdspan`/`mdarray` approach standardization, because ideally these will become the idiomatic multidimensional array (view) types for C++. Since we would like to retain the traditional BLAS-like interfaces for those familiar with them, I think it would make sense to add an `mdspan`/`mdarray`-type interface in parallel with the existing one (one interface can be implemented in headers as a wrapper around the other without much work).

Ideally, P1673 will be accepted into the STL, and then, instead of defining our own interface, we can simply implement P1673.

What do other people think?

As you mentioned, [`basic_`]`mdspan` is the natural replacement for USM pointers, but `basic_mdarray` (with a `sycl::buffer` as storage) is the natural replacement for buffers. From oneMKL’s point of view, the host code’s ownership of the data isn’t crucial for implementation, so I’d say an `mdspan` interface is sufficient (and users having mdarrays can get a `.view()`).

**Comment by petercad** [comment]
Also, I wanted to mention we are hosting the oneMKL specification [as part of the oneAPI specification](https://github.com/oneapi-src/oneapi-spec), so let's move the discussion to a new issue there.

**Comment by mhoemmen** [comment]
@petercad wrote:
> ...but `basic_mdarray` (with a `sycl::buffer` as storage) is the natural replacement for buffers.

Sort of?  `sycl::buffer` only permits direct array access through an accessor.  `basic_mdarray` doesn't have an interface for getting an accessor.  Perhaps it should?  Now would be a good time to contribute to P1684 if you think it should change.

> We feel that a more elegant and feature rich C++-based BLAS proposal would require extensive coordination with and buy-in from University of Tennessee Netlib owners.

The 1-D `sycl::buffer` interface to BLAS algorithms isn't useful for implementing LAPACK algorithms, because there's no way to say "I'm only accessing this subset of the matrix."  I imagine that anyone (UTK folks included) needing to implement matrix factorizations would only ever use the interface that takes raw pointers and lets the caller specify dependencies explicitly.

The raw pointers interface is fine, but why not just make it a C interface?  I hope I'm not being too blunt by saying that I'd rather have an actual C interface with a way to declare data dependencies explicitly, than an interface that is C-like but not callable from C programs.  It reminds me a bit too much of MPI's C++ interface.

**Comment by Ruyk** [comment]
> The 1-D sycl::buffer interface to BLAS algorithms isn't useful for implementing LAPACK algorithms, because there's no way to say "I'm only accessing this subset of the matrix."

In SYCL 1.2.1, you can create sub-buffers from buffers, would that not be enough? what is missing to the sub-buffer functionality to make it happen?

**Comment by mhoemmen** [comment]
@Ruyk wrote:
> In SYCL 1.2.1, you can create sub-buffers from buffers, would that not be enough?

"Sub-buffers must be constructed from a contiguous region of memory in a buffer." -- SYCL 1.2.1 specification, Section 4.7.2, last (text) paragraph.  Matrix factorization algorithms (what the BLAS was designed to support) need to be able to address submatrices that have noncontiguous storage in general.

Giving the entire buffer to all the BLAS algorithms that make up a matrix factorization means that I lose any potential parallelism.  In SYCL terms, all the buffers are overlapping, so SYCL must schedule all the operations sequentially (see SYCL 1.2.1 spec, Section 3.5.1).  This is bad because newer matrix factorization algorithms, like those implemented by UTK's PLASMA and MAGMA libraries, have a lot of potential parallelism across submatrices.  The only way to work around this issue would be to reorganize the matrix into a different layout on entry and exit, so that submatrices are all stored contiguously.

**Comment by Ruyk** [comment]
It could be possible to add non-contiguous sub-buffers to SYCL. A SYCL implementation can implement this on top of OpenCL (as long as sub-buffers are non-overlapping) with varying degrees of success. Other backends, such as CUDA, can implement this easily. 
Would that make the buffer interface more useful?

**Comment by mhoemmen** [comment]
@Ruyk Noncontiguous sub-buffers would indeed make oneMKL's proposed buffer interface more useful.  I haven't been following the SYCL standardization process, so I don't know if there are proposals in flight to support noncontiguous sub-buffers.  Regardless, it would make sense.

**Comment by petercad** [comment]
Closing this issue for now due to lack of activity (feel free to open new issues [here](https://github.com/oneapi-src/oneapi-spec)). Internally we will soon be working on a new, more C++-idiomatic, BLAS-like interface that can be called within user-written device code, and will welcome feedback on it.

**Comment by mhoemmen** [comment]
@petercad I'm glad you're working on a more idiomatic C++ interface!  I'll be happy to help review it.  Please do consider using `basic_mdspan` as a low-level representation of matrices and vectors!

**Comment by petercad** [comment]
@mhoemmen I agree, `basic_mdspan` and a free function interface will be my first choice.

Outside of device code, I don't see any real impediments to implementing P1673 (with either buffers or USM) directly.

Inside of device code, to fit the SYCL programming model, we'll typically want multiple logical threads (-> SIMD lanes/cores) to cooperatively perform operations, so the interfaces will need to be slightly more exotic. In addition, user code will want objects with owning semantics (a la `basic_mdarray`) to hold matrix/vector data that ought to live in registers. Due to technical implementation considerations, I'm not sure yet if actual `basic_mdarray` objects (with a suitable accessor) will suffice, but hopefully they will.

We'd be glad to have your review once it's prototyped!

**Comment by mhoemmen** [comment]
@petercad wrote:
> Inside of device code, to fit the SYCL programming model, we'll typically want multiple logical threads (-> SIMD lanes/cores) to cooperatively perform operations, so the interfaces will need to be slightly more exotic.

Would P1673's `ExecutionPolicy&&` argument (where `ExecutionPolicy` is whatever you need to pass in) cover this use case?  (For instance, you could pass in something that represents a group of threads.)  If not, then please let's talk about how P1673 could change to support your needs.  (You're literally a [wg21.link/p1673](coauthor of P1673), right?)

> In addition, user code will want objects with owning semantics (a la `basic_mdarray`) to hold matrix/vector data that ought to live in registers. 

Earlier versions of P1673 aimed to cover this use case, with interfaces taking `basic_mdspan` by value or `basic_mdarray` by reference.  We took that out of the most recent version.  Do you think it's important for these functions to return the matrix or vector by value?  At previous WG21 meetings and CppCon 2019, I kept asking people that question, but didn't get a conclusive answer.
